sdk/program/src/stake/state.rs

```diff
@@ -567,13 +567,31 @@ impl Stake {
         Ok(new)
     }
 
-    pub fn deactivate(&mut self, epoch: Epoch) -> Result<(), StakeError> {
+    pub fn deactivate(
+        &mut self,
+        epoch: Epoch,
+        stake_flags: &StakeFlags,
+        history: Option<&StakeHistory>,
+    ) -> Result<(), StakeError> {
         if self.delegation.deactivation_epoch != std::u64::MAX {
-            Err(StakeError::AlreadyDeactivated)
-        } else {
-            self.delegation.deactivation_epoch = epoch;
-            Ok(())
+            return Err(StakeError::AlreadyDeactivated);
         }
+
+        // when MUST_FULLY_ACTIVATE_BEFORE_DEACTIVATION_IS_PERMITTED flag is set on stake_flags,
+        // deactivation is only permitted when the stake delegation activating amount is zero.
+        if stake_flags.contains(StakeFlags::MUST_FULLY_ACTIVATE_BEFORE_DEACTIVATION_IS_PERMITTED) {
```

Comment:
i think we're actually going to need to lift the flag check up one level so it can be explicitly feature gated.  as is, if someone were able to manipulate a stake account to set the flag before feature activation, they could induce an error priority reordering and fork the cluster.